src/printer.js

```diff
@@ -2762,62 +2763,83 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
           : util.htmlEscapeInsideAngleBracket(child.value);
         const value = partiallyEscapedValue.replace(/\u00a0/g, "&nbsp;");
 
-        if (/\S/.test(value)) {
-          // treat each line of text as its own entity
-          value.split(/(\n\s*)/).forEach(line => {
-            const newlines = line.match(/\n/g);
-            if (newlines) {
-              children.push(hardline);
-
-              // allow one extra newline
-              if (newlines.length > 1) {
-                children.push(hardline);
-              }
-              return;
-            }
 
-            const beginSpace = /^\s+/.test(line);
-            if (beginSpace) {
-              children.push(jsxWhitespace);
-              children.push(softline);
-            }
+        // // console.log('value', value);
+        // if (/\S/.test(value)) {
+        //   value.split(/(\s+)/).forEach(word => {
+        //     // const newlines = word.match(/\n/g);
+        //     // if (newlines) {
+        //     //   children.push(hardline);
+        //     //
+        //     //   // allow one extra newline
+        //     //   if (newlines.length > 1) {
+        //     //     children.push(hardline);
+        //     //   }
+        //     //   return;
+        //     // }
+        //
+        //     const space = /^\s+$/.test(word);
+        //     if (space) {
+        //       words.push(line);
+        //       return;
+        //     }
+        //
+        //     const stripped = word.replace(/^\s+|\s+$/g, "");
+        //     if (stripped) {
+        //       words.push(stripped);
+        //     }
+        //   });
+        //
+        //   if (!isLineNext(util.getLast(words))) {
+        //     words.push(softline);
+        //   }
+        // } else if (/\n/.test(value)) {
+        //   words.push(hardline);
+        //
+        //   // allow one extra newline
+        //   if (value.match(/\n/g).length > 1) {
+        //     words.push(hardline);
+        //   }
+        // } else if (/\s/.test(value)) {
+        //   // whitespace-only without newlines,
+        //   // eg; a single space separating two elements
+        //   words.push(jsxWhitespace);
+        //   words.push(softline);
+        // }
+
+        value.replace(/^\s+|\s+$/g, "").split(/(\s+)/).forEach(word => {
```

Comment:
I don't really have an intent at the moment, just playing around with what is possible 😀

Currently it is preserving new lines **within** JSX text, but I expect it may compact the join between text and tags.

I was hoping after a period of playing around we'd know enough about the possibilities to come up with a coherent set of layout rules that could then be implemented.